python/src/scipp/compat/mantid.py

```diff
@@ -195,15 +195,70 @@ def validate_and_get_unit(unit):
         return known_units[unit]
 
 
-def init_pos(ws):
-    nHist = ws.getNumberHistograms()
-    pos = np.zeros([nHist, 3])
+def _to_spherical(x, y, z):
+    r = np.sqrt((x * x) + (y * y) + (z * z))
+    theta = np.arccos(z / r)
+    phi = np.arctan2(y, x)
+    return r, theta, phi
+
+
+def _to_cartesian(r, theta, phi):
+    x = r * np.sin(theta) * np.cos(phi)
+    y = r * np.sin(theta) * np.sin(phi)
+    z = r * np.cos(theta)
+    return x, y, z
+
+
+def rotation_matrix_from_vectors(source, target):
+    """ Find the rotation matrix that aligns source to target
+    :param source: A 3d "source" vector
+    :param target: A 3d "destination" vector
+    :return mat: A  matrix (3x3) to transform source to target.
+    """
+    a, b = (source / np.linalg.norm(source)).reshape(3), (
+        target / np.linalg.norm(target)).reshape(3)
+    v = np.cross(a, b)
+    c = np.dot(a, b)  # cos
+    s = np.linalg.norm(v)  # sin
+    if s == 0:
+        return np.eye(3)
+    kmat = np.array([[0, -v[2], v[1]], [v[2], 0, -v[0]], [-v[1], v[0], 0]])
+    rotation_matrix = np.eye(3) + kmat + kmat.dot(kmat) * ((1 - c) / (s**2))
+    return rotation_matrix
+
+
+def init_pos(ws, source_pos, sample_pos):
+    import numpy as np
+
+    if sample_pos and source_pos:
+        act_beam = (sample_pos - source_pos).values
+        rot = rotation_matrix_from_vectors(act_beam, [0, 0, 1])
+        inv_rot = rot.transpose()
+    else:
+        rot = np.eye(3)
+        inv_rot = rot
 
     spec_info = ws.spectrumInfo()
-    for i in range(nHist):
+    det_info = ws.detectorInfo()
+    pos = np.zeros([len(spec_info), 3])
+    for i in range(len(spec_info)):
         if spec_info.hasDetectors(i):
-            p = spec_info.position(i)
-            pos[i, :] = [p.X(), p.Y(), p.Z()]
+            definition = spec_info.getSpectrumDefinition(i)
+            theta, phi, r = 0.0, 0.0, 0.0
+            n_dets = len(definition)
```

Comment:
I am a bit worried about the performance implications. Can we just skip all the complicated bits in case there is just a single detector per spectrum? Have you checked the performance?